p4-fusion/main.cc

```diff
@@ -63,9 +64,77 @@ int Main(int argc, char** argv)
 		return 1;
 	}
 
-	// Set the signal here because it gets reset after P4API library is initialized.
-	std::signal(SIGINT, SignalHandler);
-	std::signal(SIGTERM, SignalHandler);
+	// Block signals from being handled by the main thread, and all future threads.
+	//
+	// - SIGINT, SIGTERM, SIGHUP: thread pool will be shutdown and std::exit will be called
+	// - SIGUSR1: only sent by main() to tell the signal handler thread to exit
+	sigset_t blockedSignals;
+	sigemptyset(&blockedSignals);
+	for (auto sig : { SIGINT, SIGTERM, SIGHUP, SIGUSR1 })
+	{
+		sigaddset(&blockedSignals, sig);
+	}
+
+	int rc = pthread_sigmask(SIG_BLOCK, &blockedSignals, nullptr);
+	if (rc != 0)
+	{
+		ERR("(signal handler) failed to block signals: (" << errno << ") " << strerror(errno));
+		return 1;
+	}
+
+	// Create the thread pool
+	int networkThreads = arguments.GetNetworkThreads();
+	PRINT("Network Threads: " << networkThreads);
+	PRINT("Creating " << networkThreads << " network threads");
+	ThreadPool pool(networkThreads);
+	SUCCESS("Created " << pool.GetThreadCount() << " threads in thread pool");
+
+	// Spawn a thread to handle signals.
+	// The thread will block and wait for signals to arrive and then shutdown the thread pool, unless it receives SIGUSR1,
+	// in which case it will just exit (since main() is handling the shutdown).
+	//
+	// Using a separate thread for purely signal handling allows us to use non-reentrant functions
+	// (such as std::cout, condition variables, etc.) in the signal handler.
+	auto signalHandlingThread = std::thread(
+	    [&blockedSignals, &pool]()
+	    {
+		    auto shutdown = [&pool]()
+		    {
+			    pool.ShutDown();
+
+			    P4API::ShutdownLibraries();
+
+			    mtr_flush();
+			    mtr_shutdown();
+		    };
+
+		    // Wait for signals to arrive.
+		    int sig = 0;
+		    int rc = sigwait(&blockedSignals, &sig);
```

Comment:
This is quite confusing to read because of the variable name `blockedSignals`, because in this sigwait call, it's not a "set by negation" but by inclusion. Moreover, it is being captured by reference inside this thread, which creates a concern around race conditions/concurrent mutation.

I recommend instead creating a copy of that variable into a separate one while capturing it by value. Example code: https://godbolt.org/z/8sav5vM37

Reply:
Sure, happy to change the name and copy it. Done in [`9cf654e` (#9)](https://github.com/sourcegraph/p4-fusion/pull/9/commits/9cf654e4ea61b9973b0f90cd6986d2ae46c88e9a)

(Unfortunately, I had to use slightly uglier syntax since we can't initialize variables for closures in C++11.)